File: Go/tema2/tema2.go
package tema2

import (
	"bufio"
	"fmt"
	"log"
	"net/http"
	"os"
	//"os/exec"
	//"runtime"
	"strconv"
	"strings"
	"time"
	"Go/grafico"
	"Go/tabla"

	"github.com/go-rod/rod"
	"github.com/gocolly/colly/v2"
)

var CANTIDAD_BARRAS int = 20

type Resultados struct {
	horaUltimaAct string
	listaTopics   []string
}

func Run(nombreaArchivo1, directorioArchivo1, nombreArchivo2, directorioArchivo2 string) {
	fmt.Println("Empieza el programa...")
	localFileName := "data.html"
	link := "https://github.com/topics/nodejs?o=desc&s=updated"
	cantRepeticiones := 33

	fmt.Println("Generando Archivo.\nEste proceso puede tomar un tiempo...")

	// Generamos el archivo local con todos los articulos
	generandoArchivoHTML(localFileName, link, cantRepeticiones)

	fmt.Println("Teminado el proceso de generacion de archivo!")

	var listaArticulos []Resultados

	listaArticulos = getHorasTopics(localFileName)

	// Guardamos los resultados en un archivo
	guardarListaResultados(directorioArchivo1 + "/" + nombreaArchivo1, listaArticulos)

	// Eliminamos los resultados que tienen mas de 30 dias de antigüedad
	listaArticulos = limpiarLista(listaArticulos)

	fmt.Println("Cant. Total de Articulos: " + strconv.Itoa(len(listaArticulos)))
	var mapeo map[string]int
	mapeo = contarTopics(listaArticulos)

	// Pasando el array para ordenar
	lista := make(grafico.Resultado, 0, len(mapeo))

	for topic, apariciones := range mapeo {
		lista = append(lista, grafico.Datos{Nombre: topic, NroApariciones: apariciones})
	}

	fmt.Println("Presione ENTER para continuar...")
	bufio.NewReader(os.Stdin).ReadString('\n')
	// Ordenando
	grafico.OrdenarListaResultados(&lista)
	fmt.Println("Lista Ordenada de Resultados: ")

	// Encabezados
	var encabezado = []string {"TOPIC", "MENCIONES"}

	matrizResultados := make([][]string, 0)
	for i := 0; i < len(lista); i++ {
		nombre := lista[i].Nombre
		cantidad := fmt.Sprintf("%d", lista[i].NroApariciones)
		var fila = []string {nombre, cantidad} 
		matrizResultados = append(matrizResultados, fila)
	}
	
	tabla.GenerarTabla(encabezado, matrizResultados)
	
	verificarExisteDirectorio(directorioArchivo2)
	fmt.Println("A continuación, se genera el archivo del gráfico de barras.")

	texto := grafico.Info{
		Titulo: "Temas Relacionados con Nodejs",
		Subtitulo: "Los 20 primeros temas, de mayor a menor",
		EjeX: "NOMBRE",
		EjeY: "Nº de Apariciones"}

	grafico.GenerarGraficoBarras(lista, directorioArchivo2, nombreArchivo2, CANTIDAD_BARRAS, texto)
	fmt.Println("Fin del programa")
}

// Verifica si existe el directorio "dir"
// Si no existe, lo crea
// Si ya existe, no hace nada
func verificarExisteDirectorio(dir string) {
	if _, err := os.Stat(dir); os.IsNotExist(err) {
		if err := os.Mkdir(dir, os.ModePerm); err != nil {
			log.Fatal(err)
		}
	}
}

// Contamos el nro de apariciones de cada topic.
// Retornamos los resultados en un mapeo
func contarTopics(lista []Resultados) map[string]int {
	var aux map[string]int
	aux = make(map[string]int)

	for i := 0; i < len(lista); i++ {
		for j := 0; j < len(lista[i].listaTopics); j++ {
			// Si el texto aun no existe en el mapeo
			if aux[lista[i].listaTopics[j]] == 0 {
				aux[lista[i].listaTopics[j]] = 1
			} else { // Si el texto ya existe en el mapeo
				aux[lista[i].listaTopics[j]]++
			}
		}
	}
	return aux
}

//
func getHorasTopics(link string) []Resultados {

	var listaRetorno []Resultados

	c := colly.NewCollector()
	c.OnHTML("article.border.rounded.color-shadow-small.color-bg-subtle.my-4", func(e *colly.HTMLElement) {
		r := Resultados{horaUltimaAct: strings.TrimSpace(e.ChildAttr("relative-time.no-wrap", "datetime"))}
		listaTopics := e.ChildTexts("a.topic-tag.topic-tag-link.f6.mb-2")

		if listaTopics != nil {
			for i := 0; i < len(listaTopics); i++ {
				listaTopics[i] = strings.TrimSpace(listaTopics[i])
			}
			r.listaTopics = listaTopics
			listaRetorno = append(listaRetorno, r)
		}

	})

	// Creamos un protocolo de transporte para manipular archivos locales

	t := &http.Transport{}
	t.RegisterProtocol("file", http.NewFileTransport(http.Dir("/")))
	d := &http.Client{Transport: t}
	c.WithTransport(d.Transport)
	path, err := os.Getwd()
	if err != nil {
		log.Println(err)
	}
	fmt.Println(path)
	fmt.Println("Nuevo Path: ")
	pathArchivoLocal := "file://" + pathLocal(path) + link
	c.Visit(pathArchivoLocal)
	return listaRetorno
}

// Cambia los "\" por "/" y borra mencion del disco actual
func pathLocal(oldPath string) string {
	s := strings.ReplaceAll(oldPath, "\\", "/")
	return s[strings.Index(s, ":")+1:] + "/"
}

// Accede a la pagina web, realiza el proceso "Load More"
// de manera consecutivas y retorna el archivo .html generado
func generandoArchivoHTML(nombreArchivo string, link string, cantidad int) {

	page := rod.New().
		MustConnect().
		MustPage(link)

	// Esperamos a que toda la pagina se haya cargado
	page.MustElement(`body > footer`).MustWaitVisible()

	// Realizamos los clicks consecutivos al boton "Load More"
	for i := 0; i < cantidad; i++ {
		fmt.Printf("...")
		page.MustElement(`button.ajax-pagination-btn.btn.btn-outline.color-border-default.f6.mt-0.width-full`).MustClick()
		page.MustElement(`body > footer`).MustWaitVisible()
		time.Sleep(3 * time.Second)
	}

	f, err := os.Create(nombreArchivo)
	// Error
	if err != nil {
		log.Println(err)
	}
	// Recuperamos la pagina resultante completa
	data, _ := page.HTML()
	// Escribimos en el archivo
	f.WriteString(data)

	f.Close()
}

// Guarda los resultados de la lista en un archivo .csv
func guardarListaResultados(direccionNombre string, lista []Resultados) {
	archivo, err := os.Create(direccionNombre)

	if err != nil {
		fmt.Println(err)
	}

	// Concatenamos cada lista y escribimos en el archivo
	for i := 0; i < len(lista); i++ {
		aux := strings.Join(lista[i].listaTopics, ",")
		archivo.WriteString(aux)
	}

	archivo.Close()
}

// Eliminamos los resultados que tienen mas de 30 dias de antigüedad
func limpiarLista(lista []Resultados) []Resultados {
	listaNueva := make([]Resultados, 0)

	// Tiempo Actual y de Hace Mes
	tiempoActual := time.Now().UTC()
	tiempoHaceUnMes := tiempoActual.AddDate(0, -1, 0)

	for i := 0; i < len(lista); i++ {
		tiempoArticulo, _ := time.Parse(time.RFC3339, lista[i].horaUltimaAct)
		if tiempoArticulo.After(tiempoHaceUnMes) == true {
			var aux Resultados
			aux.horaUltimaAct = lista[i].horaUltimaAct
			aux.listaTopics = lista[i].listaTopics
			listaNueva = append(listaNueva, aux)
		}
	}

	return listaNueva
}

File: Java/src/main/java/tema2/Tema2.java
package tema2;


import com.gargoylesoftware.htmlunit.WebClient;
import com.gargoylesoftware.htmlunit.html.DomNode;
import com.gargoylesoftware.htmlunit.html.DomNodeList;
import com.gargoylesoftware.htmlunit.html.HtmlPage;
import java.io.IOException;
import java.util.ArrayList;
import java.util.Collections;
import java.util.Comparator;
import java.util.logging.Level;
import java.util.logging.Logger;
import lib.Grafico;
import lib.Tabla;
import org.w3c.dom.Node;
import tema1.Tema1;

import java.time.LocalDateTime;



/**
 *
 *
 */
public class Tema2 {

    private class Resultados {
        String horaUltimaAct = "";
        ArrayList<String> listaTopics = new ArrayList(); 
    }
    
    private class Registro {
        String nombreTopic;
        Integer cantidad;
        
        Registro(String nombreTopic, Integer cantidad) {
            this.nombreTopic = nombreTopic;
            this.cantidad = cantidad;
        }
    }
    
    private int cantRepeticiones = 33;
    
    public void run(String nombreArchivo1, String directorioArchivo1, String nombreArchivo2, String directorioArchivo2) {
        System.out.println("Iniciando Programa 2...");
        String localFileName = "data.html";
        
        // Para analizar 
//        String link = "https://github.com/topics/nodejs?o=desc&s=updated";
        String link = "https://github.com/topics/python?o=desc&s=updated";
        System.out.println("Procesando.\nEste proceso puede tomar un tiempo...");
        
        // Generando el archivo local con los articulos
        ArrayList<Resultados> listaArticulos = procesar(link);

        System.out.println("Cantidad de Articulos: " + listaArticulos.size());

        // Creamos una lista de los topics y cantidad de apariciones
        ArrayList<Registro> listaResultados = cargarListaRegistro(listaArticulos);
        
        // Ordenamos la lista
        Collections.sort(listaResultados, new Comparator<Registro>() {
            @Override
            public int compare(Registro reg1, Registro reg2) {
            return reg1.cantidad > reg2.cantidad ? -1 : (reg1.cantidad < reg2.cantidad) ? 1 : 0;
        }});

        // Encabezados        
        String[] encabezado = {"TOPIC", "MENCIONES"};
        
        // Datos de la tabla
        String[][] matrizResultados = new String[listaResultados.size()][encabezado.length];        
	for (int i = 0; i < listaResultados.size(); i++) {
            matrizResultados[i][0] = listaResultados.get(i).nombreTopic;
            matrizResultados[i][1] = listaResultados.get(i).cantidad.toString();
	}
        Tabla t = new Tabla();
        t.generarTabla(encabezado, matrizResultados);
        
        // Generando los datos para el grafico de barras de los 20
        // temas con mayor cantidad de menciones.
        String[] info = {"Temas relacionados con Nodejs", "Los 20 topics mas mencionados", "TOPIC", "MENCIONES"};
        
        String[][] datos = new String[20][3];
        for(int i = 0; i < datos.length; i++) {
            datos[i][0] = listaResultados.get(i).cantidad.toString();
            datos[i][1] = listaResultados.get(i).nombreTopic;
            datos[i][2] = "";
        }
        Grafico g = new Grafico(info, datos);
        g.Run();
        
    }
    
    /***
     * Recibe una lista Resultados y retorna el nombre del topic y la cantidad
     * de apariciones en una lista de Registros
     */
    private ArrayList<Registro> cargarListaRegistro(ArrayList<Resultados> lista) {
        ArrayList<Registro> ret = new ArrayList<Registro>();
            
        for(int i = 0; i < lista.size(); i++) {
            for(int j = 0; j < lista.get(i).listaTopics.size(); j++) {
                boolean existe = false;
                int pos = -1;
                String auxTopic = lista.get(i).listaTopics.get(j);
                String aaaaa = auxTopic;
                for(int k = 0; k < ret.size(); k++) {
                    if(ret.get(k).nombreTopic.equals(auxTopic)) {
                        existe = true;
                        Integer cant = ret.get(k).cantidad + 1;
                        ret.set(k, new Registro(auxTopic, cant));
                        break;
                    }
                }
                if(existe==false) {
                    Registro a = new Registro(auxTopic, 1);
                    ret.add(a);
                }
            }
        }
        return ret;
    }
    
    /***
     * Visitando la web y extrayendo los datos en una lista de Resultados
     */
    private ArrayList<Resultados> procesar(String link) {
        ArrayList<Resultados> lista = new ArrayList<Resultados>();
        boolean continuar = true;
        
        // Creamos el objeto con el tiempo actual.
        LocalDateTime tiempoActual = LocalDateTime.now();
        // Definimos el tiempo que sirve de limite (hace 30 dias)
        LocalDateTime tiempoLimite = tiempoActual.plusDays(-30);
        
        continuar = procesarPagina(lista, link, tiempoLimite);
        for(int i = 2; i < cantRepeticiones && continuar; i++) {
            System.out.printf("... ");
            continuar = procesarPagina(lista, link + "&page=" + Integer.toString(i), tiempoLimite);
            try {
                Thread.sleep(10000);
            } catch (InterruptedException ex) {
                Logger.getLogger(Tema2.class.getName()).log(Level.SEVERE, null, ex);
            }
        }
        System.out.println("");
        return lista;
    }
    
    private boolean procesarPagina (ArrayList<Resultados> lista, String link, LocalDateTime tiempoLimite) {
        boolean continuar = true;

        // Iniciamos el cliente web
        WebClient webClient = new WebClient();
        
        // remove log
        java.util.logging.Logger.getLogger("com.gargoylesoftware").setLevel(Level.OFF); 
        System.setProperty("org.apache.commons.logging.Log", "org.apache.commons.logging.impl.NoOpLog");
        
        //configuring options    
        webClient.getOptions().setUseInsecureSSL(true);
        webClient.getOptions().setCssEnabled(true);
        webClient.getOptions().setThrowExceptionOnFailingStatusCode(false);
        webClient.getOptions().setThrowExceptionOnScriptError(false);
        webClient.getOptions().setJavaScriptEnabled(true);
        HtmlPage page = null;
        try {
            page = webClient.getPage(link);
            
        } catch (IOException ex) {
            Logger.getLogger(Tema1.class.getName()).log(Level.SEVERE, null, ex);
        }

        // Hacemos una lista de los nodos de articulos extraidos
        DomNodeList<DomNode> listaNodos = page.querySelectorAll("article.border.rounded.color-shadow-small.color-bg-subtle.my-4");
        
        // Extramos los datos (hora de actualizacion y topics) de cada
        // articulo extraido.
        for(int i = 0; i < listaNodos.size(); i++) {
            Resultados registro = new Resultados();            
            // Extraemos la hora en el formato RFC3339
            DomNode aux = listaNodos.get(i).querySelector("relative-time.no-wrap");
            Node time = aux.getAttributes().getNamedItem("datetime");

            // Guardamos el registro de la hora en que se actualiza.
            registro.horaUltimaAct = time.getNodeValue();
            // Extramos la lista de topics para guardar en un string[]
            DomNodeList<DomNode> auxTopics = listaNodos.get(i).querySelectorAll("a.topic-tag.topic-tag-link.f6.mb-2");

            
            // Aca debe estar el validador de fecha
            LocalDateTime tiempoArticulo = LocalDateTime.parse(time.getNodeValue().substring(0, time.getNodeValue().length()-1));
            // Restando 4 horas obtenemos la zona horaria paraguaya
            tiempoArticulo = tiempoArticulo.plusHours(-4);
            
            
            // Si la lista estaba vacia, significa que se ignora el articulo
            // porque el articulo contiene un Issue, no una actualizacion.
            if(auxTopics.size() != 0) {
                // Verificamos si el tiempo del articulo figura que sucedio 
                // antes que el tiempo limite
                if(tiempoArticulo.isBefore(tiempoLimite)) {
                    continuar = false;
                    break;
                }
                registro.listaTopics = new ArrayList<String>();
                for(int j = 0; j < auxTopics.size(); j++) {
                    DomNode aux2 = auxTopics.get(j);
                    String topic = aux2.getTextContent();
                    registro.listaTopics.add(topic.trim());
                }
                lista.add(registro);
            }
    
        }

        return continuar;
    }
    
    
    
}
